src/fat12_partition/mod.rs

```diff
@@ -0,0 +1,231 @@
+mod setup;
+
+use defmt::{error, info, Format};
+pub use setup::init;
+
+#[derive(Clone, Format)]
+pub struct Partition {
+    /// Partition Status
+    pub p_status: u8,
+    /// Start cylinder (Legacy CHS)
+    pub p_cyl_begin: u8,
+    /// Start head (Legacy CHS)
+    pub p_head_begin: u8,
+    /// Start sector (Legacy CHS)
+    pub p_sect_begin: u8,
+    /// Partition Type (DOS, Windows, BeOS, etc)
+    pub p_type: u8,
+    /// End cylinder (Legacy CHS)
+    pub p_cyl_end: u8,
+    /// End head (Legacy CHS)
+    pub p_head_end: u8,
+    /// End sector
+    pub p_sect_end: u8,
+    /// Logical block address to start of partition
+    pub p_lba: u32,
+    /// Number of sectors in partition
+    pub p_size: u32,
+}
+impl core::fmt::Debug for Partition {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Partition")
+            .field("p_type", &format_args!("{:#02x}", self.p_type))
+            .field("p_status", &format_args!("{:#02x}", self.p_status))
+            .field("p_cyl_begin", &format_args!("{:#02x}", self.p_cyl_begin))
+            .field("p_cyl_end", &format_args!("{:#02x}", self.p_cyl_end))
+            .field("p_head_begin", &format_args!("{:#02x}", self.p_head_begin))
+            .field("p_head_end", &format_args!("{:#02x}", self.p_head_end))
+            .field("p_sect_begin", &format_args!("{:#02x}", self.p_sect_begin))
+            .field("p_sect_end", &format_args!("{:#02x}", self.p_sect_end))
+            .field("p_lba", &format_args!("{:#08x}", self.p_lba))
+            .field("p_size", &format_args!("{:#08x}", self.p_size))
+            .finish()
+    }
+}
+pub struct ByteReader<'a> {
+    data: &'a [u8],
+    position: u64,
+}
+impl<'a> ByteReader<'a> {
+    fn new(data: &'a [u8], position: u64) -> Self {
+        Self { data, position }
+    }
+    fn read1(&mut self) -> u8 {
+        let value = self.data[self.position as usize];
+        self.position += 1;
+        value
+    }
+    fn read4(&mut self) -> u32 {
+        let value = ((self.data[(self.position + 3) as usize] as u32) << 24)
+            + ((self.data[(self.position + 2) as usize] as u32) << 16)
+            + ((self.data[(self.position + 1) as usize] as u32) << 8)
+            + (self.data[self.position as usize] as u32);
```

Comment:
Here we could use [`from_le_bytes`](https://doc.rust-lang.org/std/primitive.u32.html#method.from_le_bytes) on a slice, something like:

```rust
let position = self.position as usize;
let slice = self.data[position .. position + 4];
let value = u32::from_le_bytes(slice.try_into().unwrap());
```